CopilotKit/packages/react-core/src/lib/copilot-task.ts

```diff
@@ -0,0 +1,136 @@
+import { AnnotatedFunction, FunctionCall, Message } from "@copilotkit/shared";
+import { CopilotContextParams } from "../context";
+import { defaultCopilotContextCategories } from "../components";
+import { fetchAndDecodeChatCompletion } from "../utils/fetch-chat-completion";
+
+export interface CopilotTaskConfig {
+  /**
+   * The instructions to be given to the assistant.
+   */
+  instructions: string;
+  /**
+   * The data to use for the task.
+   */
+  data?: any;
+  /**
+   * Function definitions to be sent to the API.
+   */
+  functions?: AnnotatedFunction<any[]>[];
+  /**
+   * Whether to include the copilot readable context in the task.
+   */
+  includeCopilotReadable?: boolean;
+
+  /**
+   * Whether to include functions defined via useMakeCopilotActionable in the task.
+   */
+  includeCopilotActionable?: boolean;
+}
+
+export class CopilotTask {
+  private instructions: string;
+  private data?: any;
+  private functions: AnnotatedFunction<any[]>[];
+  private includeCopilotReadable: boolean;
+  private includeCopilotActionable: boolean;
+
+  constructor(config: CopilotTaskConfig) {
+    this.instructions = config.instructions;
+    this.data = config.data;
+    this.functions = config.functions || [];
+    this.includeCopilotReadable = config.includeCopilotReadable || true;
+    this.includeCopilotActionable = config.includeCopilotActionable || true;
+  }
+
+  async run(context: CopilotContextParams): Promise<void> {
+    const entryPoints = this.includeCopilotActionable ? Object.assign({}, context.entryPoints) : {};
+
+    // merge functions into entry points
+    for (const fn of this.functions) {
+      entryPoints[fn.name] = fn;
+    }
+
+    let contextString = "";
+
+    if (this.data) {
+      contextString =
+        (typeof this.data === "string" ? this.data : JSON.stringify(this.data)) + "\n\n";
+    }
+
```

Comment:
did you have a specific use-case in mind for data?

it feels like something that may be useful, but if we don't have a specific use-case yet, it may be premature.

A few thoughts:
- arguably data should be typed (by making the interface generic)
- perhaps data should be provided upon running the task, rather than upon upon creating the task?
- how does the data relate to all the useMakeCopilotReadable context?
- how is data different from context provided inside the instructions string?

Reply:
Hey Atai,

Yes, there are two use cases I have in mind for `data`:

1. When `data` comes from another API call, it does not fit into our `useMakeCopilotReadable` paradigm.
2. Sometimes `data` might even come from a user input field, but it's only needed to run a `CopilotTask`. So making it "readable" temporarily seems fragile and unpredictable, although it will work.

- _arguably data should be typed (by making the interface generic)_
In this design, the type would not be used again afterwards. But maybe when data is provided when running the as per your next point there would be a small benefit in terms of type safety/ catching bugs.
- _perhaps data should be provided upon running the task, rather than upon upon creating the task?_
Good Point!
- _how does the data relate to all the useMakeCopilotReadable context?_
It's merged with the context, unless `includeCopilotReadable` is set to false
- _how is data different from context provided inside the instructions string?_
`instructions` is what to do, `data` is the context to use. But again, it makes much more sense to provide the data when running the task, as per your second point!

Reply:
Makes sense!

So should we switch it to a typed object provided on run?

Reply:
👍  Done